File: utils/src/main/java/com/example/utils/ActivitiesManager.java
package com.example.utils;

import android.app.Activity;
import android.content.Context;

import java.util.Stack;

public class ActivitiesManager {

    public static Stack<Activity> activityStack;
    private volatile static ActivitiesManager instance;

    private ActivitiesManager() {
    }

    /**
     * 单一实例
     */
    public static ActivitiesManager getInstance() {
        if (instance == null) {
            synchronized (ActivitiesManager.class) {
                if (instance == null) {
                    instance = new ActivitiesManager();
                }
            }
        }
        return instance;
    }

    /**
     * 添加Activity到堆
     */
    public void addActivity(Activity activity) {
        if (activityStack == null) {
            activityStack = new Stack<Activity>();
        }
        activityStack.add(activity);
    }

    public void remove(Activity activity) {
        try {
            if (activityStack == null) {
                return;
            }
            if (activityStack.size() == 0) {
                return;
            }
            activityStack.remove(activity);
        } catch (Exception e) {
            LogUtil.e(e);
        }
    }

    /**
     * 获取当前Activity（堆栈中最后一个压入的）
     */
    public Activity currentActivity() {
        if (activityStack == null) {
            return null;
        }
        if (activityStack.size() == 0) {
            return null;
        }
        return activityStack.lastElement();
    }

    /**
     * 获取最后一个活着的Activity
     */
    public Activity getLastAliveActivity() {
        if (activityStack == null) {
            return null;
        }
        if (activityStack.size() == 0) {
            return null;
        }
        for (int i = activityStack.size() - 1; i >= 0; i--) {
            if (null != activityStack.get(i) && !activityStack.get(i).isFinishing()) {
                return activityStack.get(i);
            }
        }
        return null;
    }

    public String getObjectSimpleName(Object object) {
        String name = null;
        try {
            name = object.getClass().getSimpleName();
        } catch (Exception e) {
            LogUtil.e(e);
        }
        return name;
    }

    /**
     * 结束指定的Activity
     */
    public void finishActivity(Activity activity) {
        if (activityStack == null) {
            return;
        }
        if (activityStack.size() == 0) {
            return;
        }

        if (activity != null) {
            activityStack.remove(activity);
            activity.finish();
        }
    }

    /**
     * 结束指定类名的Activity
     */
    public void finishActivity(Class<?> cls) {
        if (activityStack == null) {
            return;
        }
        if (activityStack.size() == 0) {
            return;
        }

        for (Activity activity : activityStack) {
            if (activity.getClass().equals(cls)) {
                finishActivity(activity);
            }
        }
    }

    /**
     * 结束所有Activity
     */
    public void finishAllActivity() {
        if (activityStack == null) {
            return;
        }
        if (activityStack.size() == 0) {
            return;
        }
        for (int i = 0, size = activityStack.size(); i < size; i++) {
            if (null != activityStack.get(i)) {
                activityStack.get(i).finish();
            }
        }
        activityStack.clear();
    }

    /**
     * 结束所有Activity
     */
    public void finishAllActivityByName(Activity object) {
        if (activityStack == null) {
            return;
        }
        for (int i = 0, size = activityStack.size(); i < size; i++) {
            if (null != activityStack.get(i) && activityStack.get(i) != object) {
                activityStack.get(i).finish();
            }
        }
    }

    /*** 是否包含 某一个 界面 ***/
    public boolean isExistByClass(Class object) {
        if (activityStack == null) {
            return false;
        }
        if (activityStack.size() == 0) {
            return false;
        }
        for (Activity activity : activityStack) {
            if (activity.getClass().equals(object)) {
                return true;
            }
        }
        return false;
    }

    /**
     * 彻底退出应用程序
     */
    public void AppExit(Context context) {
        try {
            finishAllActivity();
            android.app.ActivityManager activityMgr = (android.app.ActivityManager) context.getSystemService(Context.ACTIVITY_SERVICE);
            activityMgr.restartPackage(context.getPackageName());
            System.exit(0);
            android.os.Process.killProcess(android.os.Process.myPid());
        } catch (Exception e) {
            LogUtil.e(e);
        }
    }

}

File: library/utils/src/main/java/com/example/utils/ActivitiesManager.kt
package com.example.utils

import android.app.Activity
import java.util.*

class ActivitiesManager private constructor() {

    companion object {
        var activityStack: Stack<Activity>? = null

        @Volatile
        private var instance: ActivitiesManager? = null

        @JvmStatic
        fun getInstance(): ActivitiesManager {
            if (instance == null) {
                synchronized(ActivitiesManager::class.java) {
                    if (instance == null) {
                        instance =
                            ActivitiesManager()
                    }
                }
            }
            return instance!!
        }
    }

    /**
     * 添加Activity到堆
     */
    fun addActivity(activity: Activity) {
        if (activityStack == null) {
            activityStack = Stack()
        }
        activityStack!!.add(activity)
    }

    fun remove(activity: Activity) {
        activityStack?.remove(activity)
    }

    /**
     * 获取当前Activity（堆栈中最后一个压入的）
     */
    fun currentActivity(): Activity? {
        return activityStack?.lastElement()
    }

    /**
     * 获取最后一个活着的Activity
     */
    fun getLastAliveActivity(): Activity? {
        activityStack?.let {
            for (i in it.indices.reversed()) {
                if (it[i] != null && !it[i].isFinishing) {
                    return it[i]
                }
            }
        }
        return null
    }

    /**
     * 结束指定类名的Activity
     */
    fun finishActivity(cls: Class<*>) {
        activityStack?.let {
            for (activity in it) {
                if (activity.javaClass == cls) {
                    finishActivity(activity)
                }
            }
        }
    }

    /**
     * 结束指定的Activity
     */
    fun finishActivity(activity: Activity) {
        activityStack?.remove(activity)
        activity.finish()
    }

    /**
     * 结束所有Activity
     */
    fun finishAllActivity() {
        activityStack?.let {
            var i = 0
            val size = it.size
            while (i < size) {
                it[i]?.finish()
                i++
            }
            it.clear()
        }
    }

    /**
     * 结束所有Activity
     */
    fun finishAllActivityByName(activity: Activity) {
        activityStack?.let {
            var i = 0
            val size = it.size
            while (i < size) {
                if (null != it[i] && it[i] !== activity) {
                    it[i].finish()
                }
                i++
            }
        }
    }

    /*** 是否包含 某一个 界面  */
    fun isExistByClass(cls: Class<*>): Boolean {
        activityStack?.let {
            for (activity in it) {
                if (activity.javaClass == cls) {
                    return true
                }
            }
        }
        return false
    }
}